### ISSUE TITLE
IPv6 link-local addresses are valid endpoints

### ISSUE THREAD

**Issue by williamatwood41** [libreswan/libreswan#1498]
At the present time, an IPv6 link-local address (typically beginning with fe80) is not permitted as an endpoint for an IPsec Security Association in Libreswan.  This is OK for the present version of XFRM, because XFRM does not handle such addresses properly.

However, RFC 8994, Section 6.8.3.1, says that an ACP node that is supporting native IPsec
MUST use IPsec in tunnel mode [and]
MUST use local and peer link-local IPv6 addresses for encapsulation.

Once XFRM is fixed to deal properly with such endpoints, Libreswan will need to be updated to allow IPv6 link-local addresses.

This will undo the constraint imposed by update 4a936b2aad.

Please note that, because IPv6 link-local addresses are non-routable, and therefore the correct interface cannot be inferred from the address or the routing table, such addresses MUST be accompanied by an interface identifier, for example "fe80::1234:56ff:fe78:9abc%eth0".

This will require a change in the semantics of endpoint specification.  At present, the definition of "left" contains the following two sentences: "The value can also contain the interface name, which will then later be used to obtain the IP address from to fill in. For example %ppp0."  For IPv6 addresses, this is wrong in two ways:
1) An interface can have multiple, valid IPv6 addresses, (or a valid IPv4 address and one or more valid IPv6 addresses) so the required address is ambiguous;
2) The "interface" part is _required_ for the link-local address class, and its semantics are ("use this interface" rather than "use the address associated with this interface".

It is not clear to me how XFRM will ultimately specify the two parts of the required address.  However, if the Libreswan user writes the link-local address according to the example above, the address can be carried down the stack all the way to the XFRM invocation, and then split into two parts if necessary.

**Comment by cagney** [comment]
FYI,

> This will undo the constraint imposed by update 4a936b2aad81825ff05ead16593734636ff04e98.

The commit didn't impose any constraints.  It changed/fixed a hard-wired value.

**Comment by cagney** [comment]
FYI,

> the definition of "left" contains the following two sentences: "The value can also contain the interface name, which will then later be used to obtain the IP address from to fill in. For example %ppp0."

In general (there are weird quirky exceptions):

- `left` specifies the host address that an IKE SA will use when communicating with a peer
Here left=%eth0 would use that interface's host address.

- `leftsubnet=<selector>,...` which specifies the Child's traffic-selectors to be negotiated; or obtained from the peer.
If an end needs to force the interface then it will likely either be added to `leftsubnet`, or, I suspect, added as a new option.

**Comment by williamatwood41** [comment]
> FYI,
> 
> > This will undo the constraint imposed by update [4a936b2](https://github.com/libreswan/libreswan/commit/4a936b2aad81825ff05ead16593734636ff04e98).
> 
> The commit didn't impose any constraints. It changed/fixed a hard-wired value.

Yes, and both the old (fixed) value and the new (symbolic) value are part of a constraint that ensures that the address will be global, which is wrong.

> * `left` specifies the host address that an IKE SA will use when communicating with a peer
>       Here left=%eth0 would use that interface's host address.
When this was coded and/or documented, the coder/documenter was probably thinking that an interface has one address (which was true, in general, for IPv4).  In the IPv6 world, this is no longer true.  So, this use of <interface identifier> either needs to be dropped, or constrained to the IPv4-only case.

**Comment by cagney** [comment]
> When this was coded and/or documented, the coder/documenter was probably thinking that an interface has one address (which was true, in general, for IPv4). In the IPv6 world, this is no longer true.

Right.

> So, this use of either needs to be dropped, or constrained to the IPv4-only case.

It needs a rethink.

**Comment by cagney** [comment]
(when adding / removing vX tags I need to remember to add a note as to why)

I've sent an e-mail to the linux-ipsec list about this.

Meanwhile, Libreswan's code has evolved a little.  Given a config containing `left=fe80::1234:56ff:fe78:9abc%eth0` it's now pluto and not addconn that does the parsing.  So it's possible to split out the two fields.

But as was discussed, there needs to be an XFRM change.

>
